packages/block-library/src/block/edit.js

```diff
@@ -1,235 +1,163 @@
+import { useEntityBlockEditor } from '@wordpress/core-data';
+import { useCallback } from '@wordpress/element';
 import {
 	Placeholder,
 	Spinner,
 	Disabled,
-	ToolbarButton,
 	ToolbarGroup,
+	ToolbarButton,
 } from '@wordpress/components';
-import { withSelect, withDispatch } from '@wordpress/data';
 import { __ } from '@wordpress/i18n';
 import {
-	BlockControls,
 	BlockEditorProvider,
-	BlockList,
 	WritingFlow,
+	BlockList,
+	BlockControls,
 } from '@wordpress/block-editor';
-import { compose } from '@wordpress/compose';
-import { parse, serialize } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import ReusableBlockEditPanel from './edit-panel';
 
-class ReusableBlockEdit extends Component {
-	constructor( { reusableBlock } ) {
-		super( ...arguments );
-
-		this.startEditing = this.startEditing.bind( this );
-		this.stopEditing = this.stopEditing.bind( this );
-		this.setBlocks = this.setBlocks.bind( this );
-		this.setTitle = this.setTitle.bind( this );
-		this.save = this.save.bind( this );
-
-		if ( reusableBlock ) {
-			// Start in edit mode when we're working with a newly created reusable block
-			this.state = {
-				isEditing: reusableBlock.isTemporary,
-				title: reusableBlock.title,
-				blocks: parse( reusableBlock.content ),
-			};
-		} else {
-			// Start in preview mode when we're working with an existing reusable block
-			this.state = {
-				isEditing: false,
-				title: null,
-				blocks: [],
-			};
-		}
-	}
-
-	componentDidMount() {
-		if ( ! this.props.reusableBlock ) {
-			this.props.fetchReusableBlock();
-		}
-	}
-
-	componentDidUpdate( prevProps ) {
-		if (
-			prevProps.reusableBlock !== this.props.reusableBlock &&
-			this.state.title === null
-		) {
-			this.setState( {
-				title: this.props.reusableBlock.title,
-				blocks: parse( this.props.reusableBlock.content ),
+export default function ReusableBlockEdit( {
+	attributes: { ref },
+	clientId,
+	isSelected,
+} ) {
+	const recordArgs = [ 'postType', 'wp_block', ref ];
+
+	const {
+		reusableBlock,
+		hasResolved,
+		isEditing,
+		isSaving,
+		canUserUpdate,
+		settings,
+	} = useSelect(
+		( select ) => ( {
+			reusableBlock: select( 'core' ).getEditedEntityRecord(
+				...recordArgs
+			),
+			hasResolved: select( 'core' ).hasFinishedResolution(
+				'getEditedEntityRecord',
+				recordArgs
+			),
+			isSaving: select( 'core' ).isSavingEntityRecord( ...recordArgs ),
+			canUserUpdate: select( 'core' ).canUser( 'update', 'blocks', ref ),
+			isEditing: select(
+				'core/reusable-blocks'
+			).__experimentalIsEditingReusableBlock( clientId ),
+			settings: select( 'core/block-editor' ).getSettings(),
+		} ),
+		[ ref, clientId ]
+	);
+
+	const { editEntityRecord, saveEditedEntityRecord } = useDispatch( 'core' );
+	const { __experimentalSetEditingReusableBlock } = useDispatch(
+		'core/reusable-blocks'
+	);
+	const setIsEditing = useCallback(
+		( value ) => {
+			__experimentalSetEditingReusableBlock( clientId, value );
```

Comment:
Can we not use `useState()`? An action/selector/reducer for this feels very heavy.

Reply:
It does feel heavy! I went for this approach so that the createReusableBlock action could switch a newly created reusable block into an edit mode - we don't want that to happen automatically in any other case. I'm all ears for other ideas, but either way I don't think it should block this PR.